### ISSUE TITLE
Dynamic provider redirects back to login

### ISSUE THREAD

**Issue by computaserik** [DuendeSoftware/Support#1319]
**Which version of Duende IdentityServer are you using?**
Version 7
**Which version of .NET are you using?**
.NET 8
**Describe the bug**

I'm trying to implement dynamic providers, but can't make it work. I get redirected to the provider, log in and am redirected back to the callback with the user from the provider. Here I create an IdentityServer user and call HttpContext.SigninAsync. After this, I am redirected to /connect/authorize. At this point I want to be redirected back to the client application with a user, but instead I'm redirected to the IdentityServer login page. From here, I can select the provider and log in in an infinite loop. I tried to follow the implementation in the [sample](https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v7/UserInteraction/DynamicProviders/IdentityServerHost/Pages/ExternalLogin/Callback.cshtml.cs#L43), but when I debug into AuthenticateAsync in DefaultUserSession, the handler returns "No Principal.". The implementation is also similar to a working static provider in the same application. The external provider is Azure Entra in both cases.

Speaking of the sample: I tried running it locally after changing from SQLLite to SQL Server, but the login kind of fails there too. If I try to autenticate with the dynamic provider, I get redirected to the callback, but then back to the login site where the dynamic option is gone (?!). Then, if I log in with user name and password, I get sent back to the client, but the idp in the token is the dynamic provider.

Any hints on what I'm doing wrong?

**To Reproduce**

Log in with a dynamic provider

**Expected behavior**

Redirect back to the client application with a logged in user

** Actual behavior **

Redirect to the login page

**Comment by RolandGuijt** [comment]
We need more information to figure out what is going wrong. Can you please post a network trace using the browser development tools so we can see where the redirects go and which cookies are set? 
When using Chrome please check the "Preserve log" checkbox to see all the hops. In other browsers there should be something similar.

**Comment by computaserik** [comment]
Here is a trace from trying to log in. I omitted some cookies and tokens, as well as some large responses, with "[omitted]". Let me know if you need any of the omitted values.
[signin.txt](https://github.com/user-attachments/files/16191380/signin.txt)

**Comment by RolandGuijt** [comment]
Thank you for your trace file.
It seems that the callback request from the dynamic external provider isn't setting cookies. Can you please verify if `HttpContext.SigninAsync` is really called and executed correctly?

**Comment by computaserik** [comment]
Using the debugger I can see that `SigninAsync` is called and no exception is thrown. Are there other things I should check to see that it's executed "correctly"? 
Here is the callback code:
```
[SecurityHeaders]
public class DynamicProviderLoginController(IIdentityProviderStore identityProviderStore, IDynamicProviderLoginService dynamicProviderLoginService) : Controller
{
    private readonly IIdentityProviderStore _identityProviderStore = identityProviderStore;
    private readonly IDynamicProviderLoginService _dynamicProviderLoginService = dynamicProviderLoginService;

    [HttpGet]
    public async Task<IActionResult> Callback()
    {
        // sample implementation: https://github.com/DuendeSoftware/Samples/blob/main/IdentityServer/v7/UserInteraction/DynamicProviders/IdentityServerHost/Pages/ExternalLogin/Callback.cshtml.cs

        var authenticateResult = await AuthenticateResult();

        if (!authenticateResult.Succeeded)
        {
            return Redirect("/CookieMissing");
        }

        var scheme = authenticateResult.Properties?.Items["scheme"];
        var provider = scheme != null ? await _identityProviderStore.GetBySchemeAsync(scheme) : null;
        if (provider == null || !provider.Enabled || provider is not OidcProvider oidcProvider)
        {
            return BadRequest();
        }
        var externalUser = authenticateResult.Principal;

        var success = await _dynamicProviderLoginService.TrySigninAsync(authenticateResult, HttpContext, externalUser, scheme);
        if (!success) 
        {
            return BadRequest();
        }

        var returnUrl = authenticateResult.Properties?.Items["returnUrl"] ?? "~/";
        
        return Redirect(returnUrl);
    }

    private async Task<AuthenticateResult> AuthenticateResult()
    {
        var authenticateResult = await HttpContext.AuthenticateAsync(IdentityServerConstants.ExternalCookieAuthenticationScheme);
        return authenticateResult ?? Microsoft.AspNetCore.Authentication.AuthenticateResult.NoResult();
    }
}
```

```
public interface IDynamicProviderLoginService
{
    Task<bool> TrySigninAsync(AuthenticateResult authenticateResult, HttpContext httpContext, ClaimsPrincipal principal, string scheme);
}

public class DynamicProviderLoginService(UserManager<IdentityUser> userManager) : IDynamicProviderLoginService
{
    private readonly UserManager<IdentityUser> _userManager = userManager;

    public async Task<bool> TrySigninAsync(AuthenticateResult authenticateResult, HttpContext httpContext, ClaimsPrincipal principal, string scheme)
    {
        var identifier = principal.FindFirstValue("oid"); 
        if (identifier == null)
        {
            return false;
        }

        var user = await _userManager.FindByLoginAsync(scheme, identifier);
        if (user != null)
        {
            // reset the user during development/debugging. This if will be removed when dynamic providers work
            await _userManager.DeleteAsync(user);
            user = null;
        }
        if (user == null)
        {
            user = new IdentityUser(identifier);
            var claims = principal.Claims.ToList();
            claims.Remove(principal.FindFirst("sub"));
            var result = await _userManager.CreateAsync(user);
            if (!result.Succeeded)
            {
                return false;
            }

            result = await _userManager.AddLoginAsync(user, new UserLoginInfo(scheme, identifier, scheme));
            if (!result.Succeeded)
            {
                return false;
            }

            result = await _userManager.AddClaimsAsync(user, claims);
            if (!result.Succeeded)
            {
                return false;
            }
        }
        var localClaims = new List<Claim>();
        var signinProps = new AuthenticationProperties();
        
        CaptureExternalLoginContext(authenticateResult, localClaims, signinProps);
        // issue authentication cookie for user
        var isuser = new IdentityServerUser(user.Id)
        {
            DisplayName = principal.FindFirstValue("name"),
            IdentityProvider = scheme,
            AdditionalClaims = localClaims
        };

        await httpContext.SignInAsync(isuser, signinProps);

        // delete temporary cookie used during external authentication
        await httpContext.SignOutAsync(IdentityServerConstants.ExternalCookieAuthenticationScheme);

        return true;
    }

    // if the external login is OIDC-based, there are certain things we need to preserve to make logout work
    // this will be different for WS-Fed, SAML2p or other protocols
    private void CaptureExternalLoginContext(AuthenticateResult externalResult, List<Claim> localClaims, AuthenticationProperties localSignInProps)
    {
        // if the external system sent a session id claim, copy it over
        // so we can use it for single sign-out
        var sid = externalResult.Principal.Claims.FirstOrDefault(x => x.Type == JwtClaimTypes.SessionId);
        if (sid != null)
        {
            localClaims.Add(new Claim(JwtClaimTypes.SessionId, sid.Value));
        }

        // if the external provider issued an id_token, we'll keep it for signout
        var idToken = externalResult.Properties.GetTokenValue("id_token");
        if (idToken != null)
        {
            localSignInProps.StoreTokens(new[] { new AuthenticationToken { Name = "id_token", Value = idToken } });
        }
    }
}
```
If it helps, we can set up a meeting where I can show you live.

**Comment by RolandGuijt** [comment]
We've looked at your trace file and code. It all seems fine except for the fact that the cookie is not set in the callback step.
It's hard to determine what the problem might be. Unfortunately a live call isn't part of the support we're offering in this issue tracker.
If you want further help we could help out using our remote consulting offering. Please [get in touch](https://duendesoftware.com/contact) for more information about that.

**Comment by RolandGuijt** [comment]
Closing this for now. Feel free to reopen the issue if you feel we can help you more using this issue tracker.

**Comment by computaserik** [comment]
For future reference: I finally found the culprit. It was in our code, not IdentityServer, of course. This was in the startup logic, and I have no idea why. After removing it, dynamic providers work. I'll try to track down the origin and remove it unless I can find a good explanation.
```
services.Configure<SecurityStampValidatorOptions>(options =>
{
    options.ValidationInterval = TimeSpan.Zero;
});
```
Also, your dynamic provider sample is not broken. I got confused by the seemingly disappearing dynamic provider button, but what actually happened was that I was redirected to https://demo.duendesoftware.com/. The redirect was so fast that I didn't notice the URL change, and it looked like it all happened on my instance.

Thank you for your help, and apologies for making you dig through the traces!
